k2/csrc/host/fsa_util.cc

```diff
@@ -251,46 +262,63 @@ void GetArcIndexes2(const Array2<int32_t *, int32_t> &arc_map,
 
 void StringToFsa::GetSizes(Array2Size<int32_t> *fsa_size) {
   K2_CHECK_NE(fsa_size, nullptr);
-  fsa_size->size1 = fsa_size->size2 = 0;
+  fsa_size->size1 = static_cast<int32_t>(arcs_.size());
+  fsa_size->size2 = num_arcs_;
+}
 
+void StringToFsa::ReadArcsFromString() {
   static constexpr const char *kDelim = " \t";
   std::istringstream is(s_);
   std::string line;
   bool finished = false;  // when the final state is read, set it to true.
-  int32_t num_arcs = 0;
+  num_arcs_ = 0;
   while (std::getline(is, line)) {
     std::vector<std::string> splits;
     SplitStringToVector(line, kDelim, &splits);
     if (splits.empty()) continue;
 
     K2_CHECK_EQ(finished, false);
 
-    auto fields = StringVectorToIntVector(splits);
+    auto fields = StringVectorToFloatVector(splits);
     auto num_fields = fields.size();
-    if (num_fields == 3u) {
+    if (num_fields == 3u || num_fields == 4u) {
       Arc arc{};
-      arc.src_state = fields[0];
-      arc.dest_state = fields[1];
-      arc.label = fields[2];
+      arc.src_state = static_cast<int32_t>(fields[0]);
+      arc.dest_state = static_cast<int32_t>(fields[1]);
+      arc.label = static_cast<int32_t>(fields[2]);
+      if (num_fields == 4u) {
+        arc.weight = fields[3];
+      }
 
       auto new_size = std::max(arc.src_state, arc.dest_state);
       if (new_size >= arcs_.size()) arcs_.resize(new_size + 1);
 
       arcs_[arc.src_state].push_back(arc);
-      ++num_arcs;
-    } else if (num_fields == 1u) {
+      ++num_arcs_;
+    } else if (num_fields == 1u || num_fields == 2u) {
+      // TODO(guoguo): the original code only supports one final state. Is it
+      //               necessary to supports multiple final states?
       finished = true;
-      K2_CHECK_EQ(fields[0] + 1, static_cast<int32_t>(arcs_.size()));
+      K2_CHECK_EQ(static_cast<int32_t>(fields[0]) + 1,
```

Comment:
According to https://github.com/k2-fsa/k2/issues/178#issue-708879692

> The final-state is of course treated differently. The format will be like OpenFST, as just:
final_state
(with no cost, since we don't support a cost on the final state, there should be an arc with -1 going to final state, which
has the costs on).

I think there is only one field for the final state.

Reply:
It will make things easier if there's only one field. I thought we were going to support more general OpenFst style WFSAs.